starlette/middleware/base.py

```diff
@@ -52,17 +54,22 @@ async def body_stream() -> typing.AsyncGenerator[bytes, None]:
                         assert message["type"] == "http.response.body"
                         yield message.get("body", b"")
 
-            response = StreamingResponse(
+            nonlocal call_next_response
+
+            call_next_response = StreamingResponse(
                 status_code=message["status"], content=body_stream()
             )
-            response.raw_headers = message["headers"]
-            return response
+            call_next_response.raw_headers = message["headers"]
+            return call_next_response
 
         async with anyio.create_task_group() as task_group:
             request = Request(scope, receive=receive)
             response = await self.dispatch_func(request, call_next)
+            if call_next_response and response is not call_next_response:
+                async with recv_stream:
+                    async for _ in recv_stream:
+                        ...  # pragma: no cover
             await response(scope, receive, send)
```

Comment:
It seems to me that we will wait for `app` to make all of its calls to `send_stream.send` before actually processing the response returned by `dispatch_func`. If that is the case, and given that all the messages from `app` are being discarded, I think it would be preferable, either to call `await response(scope, receive, send)` before draining `recv_stream`, or to put draining of `recv_stream` in another task of the task group.

Of course, I do not know the relevant technologies (ASGI/the server implementations/Starlette) as well as you do, so there could very well be a reason that either my analysis is wrong, or that my suggestions are unworkable.

```suggestion
            if call_next_response and response is not call_next_response:
                async def drain_stream():
                    async with recv_stream:
                        async for _ in recv_stream:
                            ...  # pragma: no cover
                task_group.start_soon(drain_stream)
            await response(scope, receive, send)
```